From: Sebastian Ott <sebott@linux.vnet.ibm.com>
Subject: s390/eadm_sch: add support for irq statistics
Patch-mainline: v3.7-rc1
Git-commit: 2e73c2cf78f797f3ff299ca39b210bceb40ab804
References: bnc#792703,FATE#314095

s390/eadm_sch: add support for irq statistics

Add support for EADM interrupt statistics in /proc/interrupts.

Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/include/asm/irq.h |    1 +
 arch/s390/kernel/irq.c      |    1 +
 drivers/s390/cio/eadm_sch.c |    3 +++
 3 files changed, 5 insertions(+)

--- a/arch/s390/include/asm/irq.h
+++ b/arch/s390/include/asm/irq.h
@@ -29,6 +29,7 @@ enum interruption_class {
 	IOINT_CLW,
 	IOINT_CTC,
 	IOINT_APB,
+	IOINT_ADM,
 	NMI_NMI,
 	NR_IRQS,
 };
--- a/arch/s390/kernel/irq.c
+++ b/arch/s390/kernel/irq.c
@@ -54,6 +54,7 @@ static const struct irq_class intrclass_
 	{.name = "CLW", .desc = "[I/O] CLAW" },
 	{.name = "CTC", .desc = "[I/O] CTC" },
 	{.name = "APB", .desc = "[I/O] AP Bus" },
+	{.name = "ADM", .desc = "[I/O] EADM Subchannel" },
 	{.name = "NMI", .desc = "[NMI] Machine Check" },
 };
 
--- a/drivers/s390/cio/eadm_sch.c
+++ b/drivers/s390/cio/eadm_sch.c
@@ -5,6 +5,7 @@
  * Author(s): Sebastian Ott <sebott@linux.vnet.ibm.com>
  */
 
+#include <linux/kernel_stat.h>
 #include <linux/workqueue.h>
 #include <linux/spinlock.h>
 #include <linux/device.h>
@@ -138,6 +139,8 @@ static void eadm_subchannel_irq(struct s
 	EADM_LOG(6, "irq");
 	EADM_LOG_HEX(6, irb, sizeof(*irb));
 
+	kstat_cpu(smp_processor_id()).irqs[IOINT_ADM]++;
+
 	if ((scsw->stctl & (SCSW_STCTL_ALERT_STATUS | SCSW_STCTL_STATUS_PEND))
 	    && scsw->eswf == 1 && irb->esw.eadm.erw.r)
 		error = -EIO;
